File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    truck_weights = deque(truck_weights); queue = deque([])
    
    cnt = 0; sum_ = 0
    while truck_weights:
        if len(queue) == bridge_length:
            sum_ -= queue.popleft()
        if len(queue) < bridge_length:
            if sum_ + truck_weights[0] <= weight:
                tmp = truck_weights.popleft()
                queue.append(tmp)
                sum_ += tmp
            else:
                queue.append(0)
        cnt += 1
    cnt += bridge_length
    
    return cnt

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.cpp
#include <string>
#include <vector>
#include <deque>
#include <numeric>
#include <iostream>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    deque<int> Truck(truck_weights.begin(), truck_weights.end());
    deque<int> Bridge;   //무게, 경과시간
    int BSum = 0;
    
    int seconds = 0;
    while (!Truck.empty())
    {        
        if (Bridge.size() == bridge_length)
        {
            BSum -= Bridge.front();
            Bridge.pop_front();
        }
        if (Bridge.size() < bridge_length)
        {
            if ((weight - BSum) >= Truck.front())
            {
                Bridge.push_back(Truck.front());
                BSum += Truck.front();
                Truck.pop_front();
            }
            else
            {
                Bridge.push_back(0);
            }
        }

        seconds++;
    }
    
    seconds += bridge_length;
    return seconds;
}